File: api/api/v1/endpoints/catalog.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from typing import List

from core.database import get_db
from core.security import get_current_admin, get_current_account
from models.catalog import StageDefinition, TagDefinition, PropertyDefinition, EventTypeDefinition
from schemas.catalog import (
    StageDefinitionCreate, StageDefinitionUpdate, StageDefinitionResponse,
    TagDefinitionCreate, TagDefinitionUpdate, TagDefinitionResponse,
    PropertyDefinitionCreate, PropertyDefinitionUpdate, PropertyDefinitionResponse,
    EventTypeDefinitionCreate, EventTypeDefinitionUpdate, EventTypeDefinitionResponse,
)

router = APIRouter(prefix="/catalog", tags=["catalog"])


# ---------------------------------------------------------------------------
# Stage definitions
# ---------------------------------------------------------------------------

@router.get("/stages", response_model=List[StageDefinitionResponse])
async def list_stages(
    _admin=Depends(get_current_admin),
    account=Depends(get_current_account),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(StageDefinition)
        .where(StageDefinition.account_id == account.id)
        .order_by(StageDefinition.sort_order, StageDefinition.name)
    )
    return result.scalars().all()


@router.post("/stages", response_model=StageDefinitionResponse, status_code=status.HTTP_201_CREATED)
async def create_stage(
    body: StageDefinitionCreate,
    _admin=Depends(get_current_admin),
    account=Depends(get_current_account),
    db: AsyncSession = Depends(get_db),
):
    existing = await db.execute(
        select(StageDefinition).where(
            StageDefinition.account_id == account.id,
            StageDefinition.name == body.name,
        )
    )
    if existing.scalars().first():
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Stage already exists")
    stage = StageDefinition(account_id=account.id, **body.model_dump())
    db.add(stage)
    await db.commit()
    await db.refresh(stage)
    return stage


@router.put("/stages/{stage_id}", response_model=StageDefinitionResponse)
async def update_stage(
    stage_id: int,
    body: StageDefinitionUpdate,
    _admin=Depends(get_current_admin),
    account=Depends(get_current_account),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(StageDefinition).where(
            StageDefinition.id == stage_id,
            StageDefinition.account_id == account.id,
        )
    )
    stage = result.scalars().first()
    if not stage:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Stage not found")
    for field, value in body.model_dump(exclude_unset=True).items():
        setattr(stage, field, value)
    await db.commit()
    await db.refresh(stage)
    return stage


@router.delete("/stages/{stage_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_stage(
    stage_id: int,
    _admin=Depends(get_current_admin),
    account=Depends(get_current_account),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(StageDefinition).where(
            StageDefinition.id == stage_id,
            StageDefinition.account_id == account.id,
        )
    )
    stage = result.scalars().first()
    if not stage:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Stage not found")
    await db.delete(stage)
    await db.commit()


# ---------------------------------------------------------------------------
# Tag definitions
# ---------------------------------------------------------------------------

@router.get("/tags", response_model=List[TagDefinitionResponse])
async def list_tags(
    _admin=Depends(get_current_admin),
    account=Depends(get_current_account),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(TagDefinition)
        .where(TagDefinition.account_id == account.id)
        .order_by(TagDefinition.name)
    )
    return result.scalars().all()


@router.post("/tags", response_model=TagDefinitionResponse, status_code=status.HTTP_201_CREATED)
async def create_tag(
    body: TagDefinitionCreate,
    _admin=Depends(get_current_admin),
    account=Depends(get_current_account),
    db: AsyncSession = Depends(get_db),
):
    existing = await db.execute(
        select(TagDefinition).where(
            TagDefinition.account_id == account.id,
            TagDefinition.name == body.name,
        )
    )
    if existing.scalars().first():
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Tag already exists")
    tag = TagDefinition(account_id=account.id, **body.model_dump())
    db.add(tag)
    await db.commit()
    await db.refresh(tag)
    return tag


@router.put("/tags/{tag_id}", response_model=TagDefinitionResponse)
async def update_tag(
    tag_id: int,
    body: TagDefinitionUpdate,
    _admin=Depends(get_current_admin),
    account=Depends(get_current_account),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(TagDefinition).where(
            TagDefinition.id == tag_id,
            TagDefinition.account_id == account.id,
        )
    )
    tag = result.scalars().first()
    if not tag:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Tag not found")
    for field, value in body.model_dump(exclude_unset=True).items():
        setattr(tag, field, value)
    await db.commit()
    await db.refresh(tag)
    return tag


@router.delete("/tags/{tag_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_tag(
    tag_id: int,
    _admin=Depends(get_current_admin),
    account=Depends(get_current_account),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(TagDefinition).where(
            TagDefinition.id == tag_id,
            TagDefinition.account_id == account.id,
        )
    )
    tag = result.scalars().first()
    if not tag:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Tag not found")
    await db.delete(tag)
    await db.commit()


# ---------------------------------------------------------------------------
# Property definitions
# ---------------------------------------------------------------------------

@router.get("/properties", response_model=List[PropertyDefinitionResponse])
async def list_properties(
    _admin=Depends(get_current_admin),
    account=Depends(get_current_account),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(PropertyDefinition)
        .where(PropertyDefinition.account_id == account.id)
        .order_by(PropertyDefinition.key)
    )
    return result.scalars().all()


@router.post("/properties", response_model=PropertyDefinitionResponse, status_code=status.HTTP_201_CREATED)
async def create_property(
    body: PropertyDefinitionCreate,
    _admin=Depends(get_current_admin),
    account=Depends(get_current_account),
    db: AsyncSession = Depends(get_db),
):
    existing = await db.execute(
        select(PropertyDefinition).where(
            PropertyDefinition.account_id == account.id,
            PropertyDefinition.key == body.key,
        )
    )
    if existing.scalars().first():
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Property key already exists")
    prop = PropertyDefinition(account_id=account.id, **body.model_dump())
    db.add(prop)
    await db.commit()
    await db.refresh(prop)
    return prop


@router.put("/properties/{prop_id}", response_model=PropertyDefinitionResponse)
async def update_property(
    prop_id: int,
    body: PropertyDefinitionUpdate,
    _admin=Depends(get_current_admin),
    account=Depends(get_current_account),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(PropertyDefinition).where(
            PropertyDefinition.id == prop_id,
            PropertyDefinition.account_id == account.id,
        )
    )
    prop = result.scalars().first()
    if not prop:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Property not found")
    for field, value in body.model_dump(exclude_unset=True).items():
        setattr(prop, field, value)
    await db.commit()
    await db.refresh(prop)
    return prop


@router.delete("/properties/{prop_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_property(
    prop_id: int,
    _admin=Depends(get_current_admin),
    account=Depends(get_current_account),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(PropertyDefinition).where(
            PropertyDefinition.id == prop_id,
            PropertyDefinition.account_id == account.id,
        )
    )
    prop = result.scalars().first()
    if not prop:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Property not found")
    await db.delete(prop)
    await db.commit()


# ---------------------------------------------------------------------------
# Event type definitions
# ---------------------------------------------------------------------------

@router.get("/event-types", response_model=List[EventTypeDefinitionResponse])
async def list_event_types(
    _admin=Depends(get_current_admin),
    account=Depends(get_current_account),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(EventTypeDefinition)
        .where(EventTypeDefinition.account_id == account.id)
        .order_by(EventTypeDefinition.name)
    )
    return result.scalars().all()


@router.post("/event-types", response_model=EventTypeDefinitionResponse, status_code=status.HTTP_201_CREATED)
async def create_event_type(
    body: EventTypeDefinitionCreate,
    _admin=Depends(get_current_admin),
    account=Depends(get_current_account),
    db: AsyncSession = Depends(get_db),
):
    existing = await db.execute(
        select(EventTypeDefinition).where(
            EventTypeDefinition.account_id == account.id,
            EventTypeDefinition.name == body.name,
        )
    )
    if existing.scalars().first():
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Event type already exists")
    et = EventTypeDefinition(account_id=account.id, **body.model_dump())
    db.add(et)
    await db.commit()
    await db.refresh(et)
    return et


@router.put("/event-types/{et_id}", response_model=EventTypeDefinitionResponse)
async def update_event_type(
    et_id: int,
    body: EventTypeDefinitionUpdate,
    _admin=Depends(get_current_admin),
    account=Depends(get_current_account),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(EventTypeDefinition).where(
            EventTypeDefinition.id == et_id,
            EventTypeDefinition.account_id == account.id,
        )
    )
    et = result.scalars().first()
    if not et:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Event type not found")
    for field, value in body.model_dump(exclude_unset=True).items():
        setattr(et, field, value)
    await db.commit()
    await db.refresh(et)
    return et


@router.delete("/event-types/{et_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_event_type(
    et_id: int,
    _admin=Depends(get_current_admin),
    account=Depends(get_current_account),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(EventTypeDefinition).where(
            EventTypeDefinition.id == et_id,
            EventTypeDefinition.account_id == account.id,
        )
    )
    et = result.scalars().first()
    if not et:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Event type not found")
    await db.delete(et)
    await db.commit()

File: admin/src/api/catalog.ts
import apiClient from './client'

export type ValueType = 'string' | 'number' | 'boolean' | 'datetime'

// ---------------------------------------------------------------------------
// Stage definitions
// ---------------------------------------------------------------------------

export interface StageDefinition {
  id: number
  account_id: number
  name: string
  description: string | null
  color: string | null
  sort_order: number
  created_at: string
}

// ---------------------------------------------------------------------------
// Tag definitions
// ---------------------------------------------------------------------------

export interface TagDefinition {
  id: number
  account_id: number
  name: string
  description: string | null
  color: string | null
  created_at: string
}

// ---------------------------------------------------------------------------
// Property definitions
// ---------------------------------------------------------------------------

export interface PropertyDefinition {
  id: number
  account_id: number
  key: string
  display_name: string | null
  description: string | null
  value_type: ValueType
  created_at: string
}

// ---------------------------------------------------------------------------
// Event type definitions
// ---------------------------------------------------------------------------

export interface EventTypeDefinition {
  id: number
  account_id: number
  name: string
  description: string | null
  payload_schema: Record<string, any> | null
  created_at: string
}

// ---------------------------------------------------------------------------
// API calls
// ---------------------------------------------------------------------------

export const catalogApi = {
  // Stages
  listStages: async (): Promise<StageDefinition[]> => {
    const r = await apiClient.get<StageDefinition[]>('/api/v1/catalog/stages')
    return r.data
  },
  createStage: async (data: { name: string; description?: string; color?: string; sort_order?: number }): Promise<StageDefinition> => {
    const r = await apiClient.post<StageDefinition>('/api/v1/catalog/stages', data)
    return r.data
  },
  updateStage: async (id: number, data: Partial<{ name: string; description: string; color: string; sort_order: number }>): Promise<StageDefinition> => {
    const r = await apiClient.put<StageDefinition>(`/api/v1/catalog/stages/${id}`, data)
    return r.data
  },
  deleteStage: async (id: number): Promise<void> => {
    await apiClient.delete(`/api/v1/catalog/stages/${id}`)
  },

  // Tags
  listTags: async (): Promise<TagDefinition[]> => {
    const r = await apiClient.get<TagDefinition[]>('/api/v1/catalog/tags')
    return r.data
  },
  createTag: async (data: { name: string; description?: string; color?: string }): Promise<TagDefinition> => {
    const r = await apiClient.post<TagDefinition>('/api/v1/catalog/tags', data)
    return r.data
  },
  updateTag: async (id: number, data: Partial<{ name: string; description: string; color: string }>): Promise<TagDefinition> => {
    const r = await apiClient.put<TagDefinition>(`/api/v1/catalog/tags/${id}`, data)
    return r.data
  },
  deleteTag: async (id: number): Promise<void> => {
    await apiClient.delete(`/api/v1/catalog/tags/${id}`)
  },

  // Properties
  listProperties: async (): Promise<PropertyDefinition[]> => {
    const r = await apiClient.get<PropertyDefinition[]>('/api/v1/catalog/properties')
    return r.data
  },
  createProperty: async (data: { key: string; display_name?: string; description?: string; value_type?: ValueType }): Promise<PropertyDefinition> => {
    const r = await apiClient.post<PropertyDefinition>('/api/v1/catalog/properties', data)
    return r.data
  },
  updateProperty: async (id: number, data: Partial<{ display_name: string; description: string; value_type: ValueType }>): Promise<PropertyDefinition> => {
    const r = await apiClient.put<PropertyDefinition>(`/api/v1/catalog/properties/${id}`, data)
    return r.data
  },
  deleteProperty: async (id: number): Promise<void> => {
    await apiClient.delete(`/api/v1/catalog/properties/${id}`)
  },

  // Event types
  listEventTypes: async (): Promise<EventTypeDefinition[]> => {
    const r = await apiClient.get<EventTypeDefinition[]>('/api/v1/catalog/event-types')
    return r.data
  },
  createEventType: async (data: { name: string; description?: string; payload_schema?: Record<string, any> }): Promise<EventTypeDefinition> => {
    const r = await apiClient.post<EventTypeDefinition>('/api/v1/catalog/event-types', data)
    return r.data
  },
  updateEventType: async (id: number, data: Partial<{ description: string; payload_schema: Record<string, any> }>): Promise<EventTypeDefinition> => {
    const r = await apiClient.put<EventTypeDefinition>(`/api/v1/catalog/event-types/${id}`, data)
    return r.data
  },
  deleteEventType: async (id: number): Promise<void> => {
    await apiClient.delete(`/api/v1/catalog/event-types/${id}`)
  },
}
